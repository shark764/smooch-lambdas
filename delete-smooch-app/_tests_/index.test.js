const { lambda: { api: { validatePlatformPermissions } } } = require('alonzo');

jest.mock('alonzo');

validatePlatformPermissions.mockReturnValue(true);

const event = {
  params: {
    'tenant-id': '66d83870-30df-4a3b-8801-59edff162034',
    id: '5e31c81640a22c000f5d7f28',
  },
  identity: {
    'user-id': '667802d8-2260-436c-958a-2ee0f71f73f0',
  },
};

const mockGetSecretValue = jest.fn()
  .mockImplementation(() => ({
    promise: () => ({
      SecretString: JSON.stringify({
        id: 'app-id',
        secret: 'SECRET',
      }),
    }),
  }));

const mockList = jest.fn()
  .mockImplementation(() => ({
    integrations: {
      length: 0,
    },
  }));

const mockDelete = jest.fn()
  .mockImplementation(() => ({
    promise: () => ({}),
  }));

const mockDeleteSmooch = jest.fn()
  .mockImplementation(() => ({
    promise: () => ({}),
  }));

const mockPutSecretValue = jest.fn()
  .mockImplementation(() => ({
  }));

const mockSmoochCore = jest.fn(() => ({
  integrations: {
    list: mockList,
  },
  apps: {
    delete: mockDeleteSmooch,
  },
}));

jest.mock('aws-sdk', () => ({
  config: {
    update: jest.fn(),
  },
  DynamoDB: {
    DocumentClient: jest.fn().mockImplementation(() => ({ delete: mockDelete })),
  },
  SecretsManager: jest.fn().mockImplementation(() => ({
    getSecretValue: mockGetSecretValue,
    putSeceretValue: mockPutSecretValue,
  })),
}));

jest.mock('smooch-core', () => mockSmoochCore);

const { handler } = require('../index');

describe('delete-smooch-app', () => {
  describe('Everthing is successful', () => {
    it('when secret-old and secret is not provided', async () => {
      mockGetSecretValue.mockImplementationOnce(() => ({
        promise: () => ({
          SecretString: JSON.stringify({
            id: 'app-id',
            secret: 'SECRET',
          }),
        }),
      }));
      mockGetSecretValue.mockImplementationOnce(() => ({
        promise: () => ({
          SecretString: JSON.stringify({
            '5e31c81640a22c000f5d7f28-id': 'id',
            '5e31c81640a22c000f5d7f28-id-old': 'id-old',
            id: 'app-id',
            secret: 'SECRET',
          }),
        }),
      }));
      const result = await handler(event);
      expect(result).toMatchSnapshot();
    });

    it('sends back status 200 when the app is deleted successfully', async () => {
      const result = await handler(event);
      expect(result).toMatchSnapshot();
    });

    describe('Walkthrough', () => {
      beforeAll(async () => {
        await handler(event);
      });
      it('passes in the correct arguments to validatePlatformPermissions', async () => {
        expect(validatePlatformPermissions.mock.calls).toMatchSnapshot();
      });

      it('passes in the correct arguments to secretsClient.getSecretValue()', async () => {
        expect(mockGetSecretValue.mock.calls).toMatchSnapshot();
      });

      it('passes in the correct arguments to SmoochCore()', async () => {
        expect(mockSmoochCore.mock.calls).toMatchSnapshot();
      });

      it('passes in the correct arguments to smooch.integrations.list()', async () => {
        expect(mockList.mock.calls).toMatchSnapshot();
      });

      it('passes in the correct arguments to docClient.delete()', async () => {
        expect(mockDelete.mock.calls).toMatchSnapshot();
      });

      it('passes in the correct arguments to smooch.apps.delete()', async () => {
        expect(mockDeleteSmooch.mock.calls).toMatchSnapshot();
      });
    });
  });
  it('sends back status 400 when there are invalid parameters', async () => {
    const mockEvent = {
      params: {
        'tenant-id': '66d83870-30df-4a3b-8801-59edff162034',
      },
      identity: {
        'user-id': '667802d8-2260-436c-958a-2ee0f71f73f0',
      },
    };
    const result = await handler(mockEvent);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 403 when there are not enough permissions', async () => {
    validatePlatformPermissions.mockReturnValueOnce(false);
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when there is an error retrieving digital channels credentials', async () => {
    mockGetSecretValue.mockRejectedValueOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when there is an error validating digital channels credentials', async () => {
    mockGetSecretValue.mockImplementationOnce(() => ({
      promise: () => ({}),
    }));
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 400 when integrations are found for the app', async () => {
    mockList.mockImplementationOnce(() => ({
      integrations: {
        length: 1,
      },
    }));
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 400 when there is an error retrieving integrations', async () => {
    mockList.mockRejectedValueOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when there is an error deleting records in DynamoDB', async () => {
    mockDelete.mockRejectedValueOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when there is an error deleting the app', async () => {
    mockDeleteSmooch.mockRejectedValueOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when there is an error deleting app keys', async () => {
    mockGetSecretValue.mockImplementationOnce(() => ({
      promise: () => ({
        SecretString: JSON.stringify({
          id: 'app-id',
          secret: 'SECRET',
        }),
      }),
    }));
    mockGetSecretValue.mockRejectedValueOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when there is an error deleting app keys (thrown by putSecretValue)', async () => {
    mockGetSecretValue.mockImplementationOnce(() => ({
      promise: () => ({
        SecretString: JSON.stringify({
          id: 'app-id',
          secret: 'SECRET',
        }),
      }),
    }));
    mockGetSecretValue.mockImplementationOnce(() => ({
      promise: () => ({
        SecretString: JSON.stringify({
          '5e31c81640a22c000f5d7f28-id': 'id',
          '5e31c81640a22c000f5d7f28-id-old': 'id-old',
          '5e31c81640a22c000f5d7f28-secret-old': 'secret-old',
          '5e31c81640a22c000f5d7f28-secret': 'secret',
          id: 'app-id',
          secret: 'SECRET',
        }),
      }),
    }));
    mockPutSecretValue.mockRejectedValueOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });
});
