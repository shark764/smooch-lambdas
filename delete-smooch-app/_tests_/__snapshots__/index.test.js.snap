// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`delete-smooch-app Everthing is successful Walkthrough passes in the correct arguments to SmoochCore() 1`] = `
Array [
  Array [
    Object {
      "keyId": "app-id",
      "scope": "account",
      "secret": "SECRET",
      "serviceUrl": "mock-smooch-api-url",
    },
  ],
  Array [
    Object {
      "keyId": "app-id",
      "scope": "account",
      "secret": "SECRET",
      "serviceUrl": "mock-smooch-api-url",
    },
  ],
  Array [
    Object {
      "keyId": "app-id",
      "scope": "account",
      "secret": "SECRET",
      "serviceUrl": "mock-smooch-api-url",
    },
  ],
]
`;

exports[`delete-smooch-app Everthing is successful Walkthrough passes in the correct arguments to docClient.delete() 1`] = `
Array [
  Array [
    Object {
      "Key": Object {
        "id": "5e31c81640a22c000f5d7f28",
        "tenant-id": "66d83870-30df-4a3b-8801-59edff162034",
      },
      "TableName": "us-east-1-dev-smooch",
    },
  ],
  Array [
    Object {
      "Key": Object {
        "id": "5e31c81640a22c000f5d7f28",
        "tenant-id": "66d83870-30df-4a3b-8801-59edff162034",
      },
      "TableName": "us-east-1-dev-smooch",
    },
  ],
  Array [
    Object {
      "Key": Object {
        "id": "5e31c81640a22c000f5d7f28",
        "tenant-id": "66d83870-30df-4a3b-8801-59edff162034",
      },
      "TableName": "us-east-1-dev-smooch",
    },
  ],
]
`;

exports[`delete-smooch-app Everthing is successful Walkthrough passes in the correct arguments to secretsClient.getSecretValue() 1`] = `
Array [
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-account",
    },
  ],
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-app",
    },
  ],
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-account",
    },
  ],
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-app",
    },
  ],
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-account",
    },
  ],
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-app",
    },
  ],
]
`;

exports[`delete-smooch-app Everthing is successful Walkthrough passes in the correct arguments to smooch.apps.delete() 1`] = `
Array [
  Array [
    "5e31c81640a22c000f5d7f28",
  ],
  Array [
    "5e31c81640a22c000f5d7f28",
  ],
  Array [
    "5e31c81640a22c000f5d7f28",
  ],
]
`;

exports[`delete-smooch-app Everthing is successful Walkthrough passes in the correct arguments to smooch.integrations.list() 1`] = `
Array [
  Array [
    Object {
      "appId": "5e31c81640a22c000f5d7f28",
    },
  ],
  Array [
    Object {
      "appId": "5e31c81640a22c000f5d7f28",
    },
  ],
  Array [
    Object {
      "appId": "5e31c81640a22c000f5d7f28",
    },
  ],
]
`;

exports[`delete-smooch-app Everthing is successful Walkthrough passes in the correct arguments to validatePlatformPermissions 1`] = `
Array [
  Array [
    Object {
      "user-id": "667802d8-2260-436c-958a-2ee0f71f73f0",
    },
    Array [
      "PLATFORM_DIGITAL_CHANNELS_APP",
    ],
  ],
  Array [
    Object {
      "user-id": "667802d8-2260-436c-958a-2ee0f71f73f0",
    },
    Array [
      "PLATFORM_DIGITAL_CHANNELS_APP",
    ],
  ],
  Array [
    Object {
      "user-id": "667802d8-2260-436c-958a-2ee0f71f73f0",
    },
    Array [
      "PLATFORM_DIGITAL_CHANNELS_APP",
    ],
  ],
]
`;

exports[`delete-smooch-app Everthing is successful sends back status 200 when the app is deleted successfully 1`] = `
Object {
  "body": Object {
    "deleted": true,
    "message": "The app with for tenant 66d83870-30df-4a3b-8801-59edff162034 and appId 5e31c81640a22c000f5d7f28 has been deleted successfully",
  },
  "status": 200,
}
`;

exports[`delete-smooch-app Everthing is successful when secret-old and secret is not provided 1`] = `
Object {
  "body": Object {
    "message": "An Error has occurred trying to delete app keys",
  },
  "status": 500,
}
`;

exports[`delete-smooch-app sends back status 400 when integrations are found for the app 1`] = `
Object {
  "body": Object {
    "message": "Integrations found for this app",
  },
  "status": 400,
}
`;

exports[`delete-smooch-app sends back status 400 when there are invalid parameters 1`] = `
Object {
  "body": Object {
    "message": "Error: invalid params value \\"id\\" is required",
  },
  "status": 400,
}
`;

exports[`delete-smooch-app sends back status 400 when there are not enough permissions 1`] = `
Object {
  "body": Object {
    "message": "Error not enough permissions",
  },
  "status": 400,
}
`;

exports[`delete-smooch-app sends back status 400 when there is an error retrieving integrations 1`] = `
Object {
  "body": Object {
    "message": "An error occured trying to retrieve integrations",
  },
  "status": 400,
}
`;

exports[`delete-smooch-app sends back status 500 when there is an error deleting app keys (thrown by putSecretValue) 1`] = `
Object {
  "body": Object {
    "message": "An Error has occurred trying to delete app keys",
  },
  "status": 500,
}
`;

exports[`delete-smooch-app sends back status 500 when there is an error deleting app keys 1`] = `
Object {
  "body": Object {
    "message": "An Error has occurred trying to delete app keys",
  },
  "status": 500,
}
`;

exports[`delete-smooch-app sends back status 500 when there is an error deleting records in DynamoDB 1`] = `
Object {
  "body": Object {
    "deleted": false,
    "message": "An Error has occurred trying to delete a record in DynamoDB",
  },
  "status": 500,
}
`;

exports[`delete-smooch-app sends back status 500 when there is an error deleting the app 1`] = `
Object {
  "body": Object {
    "message": "An Error has occurred trying to delete an app",
  },
  "status": 500,
}
`;

exports[`delete-smooch-app sends back status 500 when there is an error retrieving digital channels credentials 1`] = `
Object {
  "body": Object {
    "message": "An Error has occurred trying to retrieve digital channels credentials",
  },
  "status": 500,
}
`;

exports[`delete-smooch-app sends back status 500 when there is an error validating digital channels credentials 1`] = `
Object {
  "body": Object {
    "message": "An Error has occurred trying to validate digital channels credentials",
  },
  "status": 500,
}
`;
